### PR TITLE
Multiple ledger whitelist mechanism in p2p

### PR DESCRIPTION
**Pull request by chaychen2005**

### CODE REVIEW

File: libp2p/Host.h [hunk 1]
```diff
@@ -226,6 +226,24 @@ class Host : public Worker
 
     shared_ptr<AsioInterface> const& asioInterface() const { return m_asioInterface; }
 
+    void setGroupID2NodeList(std::map<int32_t, h512s> const& _groupID2NodeList)
```
Comment (wheatli) [inline libp2p/Host.h]: define type GROUP_ID/PROTOCOL_ID/PACKET_ID in libethcore/common.h
and change all these types used in p2p and other modules to XXX_ID

Comment (chaychen2005) [reply libp2p/Host.h]: Done.

File: libethcore/Protocol.h [hunk 2]
```diff
@@ -44,15 +48,15 @@ enum ExtraIndex
...
 };
 
-inline int16_t getGroupProtoclID(int8_t groupID, uint8_t protocolID)
+inline PROTOCOL_ID getGroupProtoclID(GROUP_ID groupID, MODULE_ID moduleID)
 {
     if (groupID < 0)
         return 0;
     else
-        return (groupID << 8) | protocolID;
+        return (groupID << 8) | moduleID;
```
Comment (wheatli) [inline libethcore/Protocol.h]: 8 should be sizeof(moduleID)

File: libp2p/Common.cpp [hunk 3]
```diff
@@ -131,8 +131,8 @@ ssize_t Message::decode(const byte* buffer, size_t size)
         return PACKET_INCOMPLETE;
     }
 
-    m_protocolID = ntohs(*((int16_t*)&buffer[4]));
-    m_packetType = ntohs(*((uint16_t*)&buffer[6]));
+    m_protocolID = ntohs(*((PROTOCOL_ID*)&buffer[4]));
```
Comment (wheatli) [inline libp2p/Common.cpp]: 4 and 6 should be dynamic

File: libethcore/Protocol.h (hunk 2 above)
Comment (chaychen2005) [reply libethcore/Protocol.h]: Done.8 should be 8*sizeof(moduleID)

File: libp2p/Common.cpp (hunk 3 above)
Comment (chaychen2005) [reply libp2p/Common.cpp]: Done.

**Pull request merged**